extensions/windowgram/init.lua

```diff
@@ -0,0 +1,178 @@
+--- === hs.windowgram ===
+---
+--- auto-layout using ascii rows similar to tmuxomatic
+
+local windowgram = {}
+
+local alert = require 'hs.alert'
+local screen = require 'hs.screen'
+
+local gridh
+local gridw
+
+--- hs.windowgram.getRects(wg)
+--- Function
+--- Converts a windowgram to a table with rects.
+---
+--- Parameters:
+---  * wg - The windowgram: A table containing strings, representing the desired
+--- window layout
+---
+--- Returns:
+---  * Table with rects which contain an X, Y, W, and H values
+---
+--- Notes:
+---  * Rects are relative to the positions in the windowgram.
+---
+--- ~~~lua
+--- local wg = {
+---   "AAAAAAAAAAAABBBBBBBBBBBB",
+---   "AAAAAAAAAAAABBBBBBBBBBBB",
+---   "AAAAAAAAAAAABBBBBBBBBBBB",
+---   "CCCCCCCCCCCCCCCCCCCCCCCC",
+---   "CCCCCCCCCCCCCCCCCCCCCCCC"
+--- }
+--- local windows = windowgram.getWindows(wg)
+--- ~~~
+---
+--- This will return the following table;
+--- ~~~lua
+--- print(wg["A"])
+--- -- prints { x=1, y=1, w=12, h=3 }
+--- print(wg["B"])
+--- -- prints { x=12, y=1, w=24, h=3 }
+--- print(wg["C"])
+--- -- prints { x=1, y=3, w=24, h=5 }
+--- ~~~
+function windowgram.getRects(wg)
+  wg = removeWhitspace(wg)
+
+  gridh = #wg
+  gridw = nil
+
+  local windows = {}
+
+  for i, line in ipairs(wg) do
+    if gridw then
+      if gridw ~= line:len() then
+        error('inconsistent grid width in windowgram')
+      end
+    else
+      gridw=line:len()
+    end
+    for column = 1, #line do
+      local char = line:sub(column, column)
+      if not windows[char] then
+        -- new window, create it with size 1x1
+        windows[char] = { x = column , y = i}
+      else
+        -- expand it
+        windows[char].w = column
+        windows[char].h = i
+      end
+    end
+  end
+  return windows
+
+end
+
+function removeWhitspace(wg)
+  local target = {}
+  for i,l in ipairs(wg) do
+    l = l:gsub("%s+","")
```

Comment:
The original code had
```lua
l = l:gsub('#.*','')        -- strip comments
l = l:gsub('%s*$','')       -- strip trailing whitespace
```

and I noticed this line from my config failing with windowgram:
```lua
  "CCCCCCCCCCCCCiiiiiiiiiii      # <-- The windowgram, it defines the shapes and positions of windows",
```